### PR TITLE
clp-s: Add support for projecting of a subset of columns during search.

### PR DESCRIPTION
**Pull request by gibber9809**
# Description
This PR adds support for basic projection during search. We allow the user to specify which columns they want to be marshalled for matching search results using the --projection command line option.

For example
`clp-s s archives '*' --archive-id XYZ --projection @timestamp level msg.error`
Would return all records in the archive XYZ, showing only the @timestamp, level, and msg.error fields.

Note that if a record doesn't contain all of columns we are asked to project we simply output the record without doing anything about the missing columns. In our example above if a record were missing all three of the fields we asked for it would simply be marshalled as `{}`.

Other things to note are that we allow projected columns to be polymorphic, we do not allow the set of projected columns to contain wildcards, and we do not allow projection to travel inside of arrays. Specifically if you have an array a containing objects with the field b then asking to project "a" is legal (leading to the whole array being marshalled), but asking for "a.b" is illegal.

Also note that unlike languages like SQL the names of columns being projected are case sensitive.

We can add support for things like case-insensitive column matching, advanced handling for records that don't contain a column we're asking for, polymorphism, and more in future PRs.

The implementation is quite straightforward. We take in a set of columns, expand them against the MPT to find all matching nodes, and then check if nodes in the schema of an ERT belong to that set when deciding whether to marshal the corresponding column.

# Validation performed
- validated that requested set of columns gets marshalled
- validated that empty JSON object is produced when matching record contains none of the requested columns
- manually validated that projected data matches expected data for small test cases
- measured significant speedup when asking for columns representing a small fraction of the data

## Release Notes

- **New Features**
	- Introduced a command-line option `--projection` for specifying projected columns.
	- Added functionality to handle projections in search operations, allowing users to define which columns to return.
	- New `Projection` class for managing and resolving column projections.

- **Bug Fixes**
	- Improved logic in JSON document handling to ensure proper closure.

- **Enhancements**
	- Updated `ArchiveReader` and `SchemaReader` to utilize projection information for schema processing.
	- Enhanced methods for accessing and managing projection columns in relevant classes.

### CODE REVIEW

**Comment by gibber9809** [comment]
Seems like `lint-check (ubuntu-latest)` is failing for reasons unrelated to this PR. It appears to be complaining about formatting in a bunch of clp and glt files.

**Comment by kirkrodrigues** [comment]
> Seems like lint-check (ubuntu-latest) is failing for reasons unrelated to this PR. It appears to be complaining about formatting in a bunch of clp and glt files.

Issue being tracked in #545. You can merge main to temporarily resolve the (new) violations.

**Review by wraymo — COMMENTED** [review]
Nice work! Do you want to point out `--projection` can only be used after all positional options or do you want to support something like `./clp-s s --projection a b c -- archive query`? Maybe we should also add a projection case in the doc later.

File: components/core/src/clp_s/search/Projection.hpp [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    public:
+        // Constructors
+        OperationFailed(ErrorCode error_code, char const* const filename, int line_number)
+                : TraceableException(error_code, filename, line_number) {}
+    };
+
+    explicit Projection(ProjectionMode mode) : m_projection_mode{mode} {}
+
+    /**
+     * Add a column to the set of columns that should be included in the projected results
```
Comment (wraymo) [inline components/core/src/clp_s/search/Projection.hpp:36]:
```suggestion
     * Adds a column to the set of columns that should be included in the projected results
```

File: components/core/src/clp_s/search/Projection.hpp [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+     * Note: we could try to generalize column resolution code/move it to the schema tree. It is
+     * probably best to write a simpler version dedicated to projection for now since types are
+     * leaf-only. The type-per-token idea solves this problem (in the absence of wildcards).
+     *
+     * @param tree
+     */
+    void resolve_columns(std::shared_ptr<SchemaTree> tree);
+
+    /**
+     * Check whether a column corresponding to given leaf node should be included in the output
```
Comment (wraymo) [inline components/core/src/clp_s/search/Projection.hpp:59]:
```suggestion
     * Checks whether a column corresponding to given leaf node should be included in the output
```

File: components/core/src/clp_s/search/Projection.hpp [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+     * @return true if the column should be included in the output, false otherwise
+     */
+    bool matches_node(int32_t node_id) const {
+        return ProjectionMode::ReturnAllColumns == m_projection_mode
+               || m_matching_nodes.contains(node_id);
+    }
+
+private:
+    /**
+     * Resolve an individual column as described by the `resolve_columns` method.
```
Comment (wraymo) [inline components/core/src/clp_s/search/Projection.hpp:70]:
```suggestion
     * Resolves an individual column as described by the `resolve_columns` method.
```

File: components/core/src/clp_s/search/Projection.hpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+        OperationFailed(ErrorCode error_code, char const* const filename, int line_number)
+                : TraceableException(error_code, filename, line_number) {}
+    };
+
+    explicit Projection(ProjectionMode mode) : m_projection_mode{mode} {}
+
+    /**
+     * Add a column to the set of columns that should be included in the projected results
+     * @param column
+     * @throws OperationFailed if column contains a wildcard
```
Comment (wraymo) [inline components/core/src/clp_s/search/Projection.hpp:38]:
```suggestion
     * @throws OperationFailed if `column` contains a wildcard
```

**Review by wraymo — DISMISSED** [review]

**Review by wraymo — APPROVED** [review]

**Pull request merged**